sale_partner_delivery_window/models/res_partner.py

```diff
@@ -0,0 +1,60 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from datetime import datetime, timedelta
+
+from odoo import _, models
+from odoo.exceptions import UserError
+from odoo.tools.date_utils import date_range
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def next_delivery_window_start_datetime(self, from_date=None):
+        """Returns the next starting datetime in a preferred delivery window.
+        If from_date is already in a delivery window, from_date is "the next"
+
+        :param from_date: Datetime object (Leave empty to use now())
+        :return: Datetime object
+        """
+        self.ensure_one()
+        if not from_date:
+            from_date = datetime.now()
+        if self.is_in_delivery_window(from_date):
+            return from_date
+        # Sort by start time to ensure the window we'll find will be the first
+        # one for the weekday
+        datetime_windows = self.get_next_windows_start_datetime(
+            from_date, from_date + timedelta(days=8)
```

Comment:
```suggestion
            from_date, from_date + timedelta(days=7)
```
date_range already add a day. So current weekday is already present at the beginning and at the end

Reply:
Why computing for a predefined range. Why not try each day after each other until you match a valid one? This could be more convenient to hook with another module that would define leave periods (holidays).

Reply:
IMO with the actual implementation, there's not much sense to compute opening times for more than a week, but I moved the number of days to function arguments so that it could be overrided as needed.
However I would strongly advise against managing holidays the same way we managed times, and we should trend to use `resource.calendar`, what will need a consequent refactoring.